force-app/main/default/classes/DMLOperations.cls

```diff
@@ -183,38 +362,111 @@ public with sharing class DMLOperations {
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
 
-		// Extract the Account Name from Contact's LastName
+        String accountName ;
 
-		// Get the account based on the name or create one if it doesn't exist
+        for (Contact contact : contactsToUpsert) {
 
-		// Link the Contact to its Account using the Account Id
+        // Get the account name based on the contact's last name
 
-		// Upsert the Contacts
-	}
+        accountName = contact.LastName;
+        
+         // Query for an existing account with the same name or create one if it doesn't exist
+
+         // Extract the Account Name from Contact's LastName
+
+        // Get the account based on the name or create one if it doesn't exist
+        List<Account> accLastNameMatch = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
 
+        if (!accLastNameMatch.isEmpty()) {
+            // Link the contact to the existing account
+            contact.AccountId = accLastNameMatch[0].Id;
+        } 
+        
```

Comment:
See if this is already being done by the upsertAccount Method. If it is, you can remove the second DML inside the loop and just stay with the one that's mentioned in the instructions.